### PR TITLE
Support Bevy v0.12.0

### LINKED ISSUE #488: Update to Bevy 0.12.0
Bevy 0.12.0 has released, please update to it.

### PR DESCRIPTION
**Pull request by divark**
## Summary
This PR makes the necessary changes in order for this library to be bevy v0.12.0 compliant. This closes #488.

## Testing
1. Clone the `bevy_ecs_tilemap` repository as-is.
2. Run all of the examples as described in the `README.md`.
3. Clone my version of `bevy_ecs_tilemap`
4. Run `cargo check` to verify there are no compilation issues.
5. Run all of the examples as described in the `README.md`.
6. Observe there are (hopefully) no differences between the original.

### CODE REVIEW

**Review by rparrett — CHANGES_REQUESTED** [review]
Nice work! This looks good to me, just a few minor notes.

@Trouv Could you let us know how this branch goes with `bevy_ecs_ldtk`?

Also, just a note for reviewers: there is some expected weirdness going on in the examples using text2d on high dpi displays due to https://github.com/bevyengine/bevy/issues/10407.

I also tested examples with the `atlas` feature and spot checked a few with webgl2.

File: examples/helpers/ldtk.rs [hunk 1]
```diff
@@ -46,14 +51,34 @@ pub struct LdtkMapBundle {
 
 pub struct LdtkLoader;
 
+#[derive(Debug, Error)]
+pub enum LdtkAssetLoaderError {
+    /// An [IO](std::io) Error
+    #[error("Could load LDTk file: {0}")]
```
[the remark this answers is not in the document]
Comment (divark) [reply examples/helpers/ldtk.rs]: Oops, thanks for catching that! Fixed.

File: examples/helpers/tiled.rs [hunk 2]
```diff
@@ -91,27 +92,35 @@ impl tiled::ResourceReader for BytesResourceReader {
 
 pub struct TiledLoader;
 
+#[derive(Debug, Error)]
+pub enum TiledAssetLoaderError {
+    /// An [IO](std::io) Error
+    #[error("Could load Tiled file: {0}")]
```
[the remark this answers is not in the document]
Comment (divark) [reply examples/helpers/tiled.rs]: Oops, I did this again, my bad. Thank you for catching this! Fixed.

File: examples/3d_iso.rs [hunk 3]
```diff
@@ -35,7 +33,7 @@ fn main() {
                 })
                 .set(ImagePlugin::default_nearest())
                 .set(AssetPlugin {
-                    watch_for_changes: ChangeWatcher::with_delay(Duration::from_millis(200)),
+                    mode: AssetMode::Unprocessed,
```
[the remark this answers is not in the document]
Comment (divark) [reply examples/3d_iso.rs]: That's true. I went ahead and removed all of these calls that I could find.

**Comment by EliottGaboreau** [comment]
Great work ! I tried testing for wasm target but stumbled on an [issue caused by a dependency](https://github.com/smol-rs/async-executor/issues/72#issuecomment-1808199962) so I'm guessing we can't test for this target right now although I don't know how necessary it would be for merging this PR

**Comment by EliottGaboreau** [comment]
Went through all the example (without the atlas feature) and I didn't find any issue !

**Review by rparrett — APPROVED** [review]
I'd like to hear back from `bevy_ecs_ldtk`, and then poke StarArawn to take a look and hopefully merge.

File: examples/helpers/ldtk.rs [hunk 4]
```diff
@@ -46,14 +51,34 @@ pub struct LdtkMapBundle {
...
+    type Error = LdtkAssetLoaderError;
+
     fn load<'a>(
         &'a self,
-        bytes: &'a [u8],
+        reader: &'a mut Reader,
+        _settings: &'a Self::Settings,
         load_context: &'a mut LoadContext,
-    ) -> BoxedFuture<'a, Result<(), anyhow::Error>> {
+    ) -> BoxedFuture<'a, Result<LdtkMap, LdtkAssetLoaderError>> {
```
Comment (rparrett) [inline examples/helpers/ldtk.rs:71]: This can be `Self::Error` (same suggestion in the `Tiled` example)

File: examples/helpers/ldtk.rs [hunk 5]
```diff
@@ -18,13 +23,13 @@ pub struct LdtkPlugin;
...
-            .add_asset_loader(LdtkLoader)
+        app.init_asset::<LdtkMap>()
+            .register_asset_loader(LdtkLoader)
             .add_systems(Update, process_loaded_tile_maps);
     }
 }
 
-#[derive(TypeUuid, TypePath)]
+#[derive(TypeUuid, TypePath, Asset)]
 #[uuid = "abf9eaf2-f21c-4b46-89b0-8aa5c42199af"]
```
Comment (rparrett) [inline examples/helpers/ldtk.rs:33]: I think that this and the `TypeUuid` derive are no longer necessary, but you might want to double check that.

**Review by rparrett — COMMENTED** [review]
Sorry, noticed a couple other minor things.

File: Cargo.toml [hunk 6]
```diff
@@ -16,25 +16,26 @@ atlas = []
 render = []
 
 [dependencies]
-bevy = { version = "0.11", default-features = false, features = [
+bevy = { version = "0.12", default-features = false, features = [
     "bevy_core_pipeline",
     "bevy_render",
     "bevy_asset",
     "bevy_sprite",
+    "multi-threaded"
```
Comment (rparrett) [inline Cargo.toml:24]: Is this also needed here? I think this might only be necessary in `dev-dependencies`.

File: examples/helpers/ldtk.rs (hunk 5 above)
Comment (divark) [reply examples/helpers/ldtk.rs:33]: `TypeUuid` and the `#uuid[...]` portions are no longer needed, but `Asset` is still needed as of testing a moment ago. I'll push changes momentarily.

File: Cargo.toml (hunk 6 above)
Comment (divark) [reply Cargo.toml:24]: Ah, it is not, good catch!

File: examples/helpers/ldtk.rs (hunk 4 above)
Comment (divark) [reply examples/helpers/ldtk.rs]: Oh, nice, that'll clean things up. Thanks!

File: examples/helpers/ldtk.rs (hunk 5 above)
Comment (divark) [reply examples/helpers/ldtk.rs]: Would you recommend I propagate these changes to other aspects of the code that utilize `TypeUuid` and `Asset` as well?

Comment (rparrett) [reply examples/helpers/ldtk.rs]: Oops, when I said "this" I meant the `#[uuid` line. Yeah, Asset is fine/needed.

And yeah, I think we can ditch the uuid for the material as well.

Comment (divark) [reply examples/helpers/ldtk.rs]: Okay, done.

**Comment by dbidwell94** [comment]
Approved, ship it :laughing:

**Review by Trouv — APPROVED** [review]
Things are looking good for `bevy_ecs_ldtk` here: https://github.com/Trouv/bevy_ecs_ldtk/pull/265

Any visual bugs I noticed aren't new (#481)

**Comment by BrookJeynes** [comment]
@divark @dbidwell94 @Trouv Hey guys, any chance we could get this merged? 😅

**Comment by dbidwell94** [comment]
Don't look at me, I'm a consumer not a maintainer 🙂

**Comment by rparrett** [comment]
> Hey guys, any chance we could get this merged? 😅

It's `StarArawn` running the show, and I've let them know this is ready to look at. Please be patient. Comments like "this branch is/isn't working great in my project" would still be helpful feedback.

**Comment by BrookJeynes** [comment]
@rparrett 

> It's StarArawn running the show, and I've let them know this is ready to look at. Please be patient.
Ah! my bad, I didn't realise you guys weren't contributors. Sorry about that.

> You don't need a crates.io release to use this.
Oh?? So I can use a "non-release" branch in my rust project? I'll take a look at that now! thanks for letting me know.

**Comment by rparrett** [comment]
> Oh?? So I can use a "non-release" branch in my rust project? I'll take a look at that now! thanks for letting me know.

Yep! Something like
```
bevy_ecs_tilemap = { git = "https://github.com/divark/bevy_ecs_tilemap", branch = "0.12-fixes" }
```

**Comment by OwlyCode** [comment]
Thank you for your work on that topic!

I migrated my project using your branch and I get this error:

```
2023-11-15T18:15:44.813550Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default
thread 'Compute Task Pool (0)' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_texture
      note: label = `texture_array`
    Dimension Z is zero
    
    ...
    
Encountered a panic in system `bevy_ecs_tilemap::render::prepare_textures`!
```

I'm not sure how I can confirm this is on my side or on the lib.

**Comment by OwlyCode** [comment]
I found the issue with my code, I didn't notice I was creating a tilemap without a texture. Which seems to works with the current version of this library.

Should this case also be handled by this PR?

**Comment by Twyker** [comment]
looking forward to this beautiful merge 🙏

**Comment by csandven** [comment]
Seems to work very well with my project as well, so another thumbs up ! Looking forward to the merge.

**Review by Trouv — CHANGES_REQUESTED** [review]
Hello again - one of my users @musjj caught a bug regarding render targets that wasn't happening in 0.11. I looked into it a little bit and it seems that, when adding a camera with a custom `RenderTarget` **after** a tilemap is rendered, we get errors like this:
```
thread '<unnamed>' panicked at $HOME/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.17.2/src/backend/direct.rs:3056:5:
wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 4, Vulkan)>`
    In a set_bind_group command
      note: bind group = `tilemap_view_bind_group`
    Dynamic binding offset index 0 with offset 2304 would overrun the buffer bound to bind group 0 -> binding 0. Buffer size is 1536 bytes, the binding binds bytes 0..608, meaning the maximum the binding can be offset is 928 bytes
```

Or on my system (similar but slightly different):
```
wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 13, Vulkan)>`
    In a set_bind_group command
      note: bind group = `<BindGroup-(11, 6, Vulkan)>`
    Dynamic binding offset index 0 with offset 608 would overrun the buffer bound to bind group 0 -> binding 0. Buffer size is 608 bytes, the binding binds bytes 0..608, meaning the maximum the binding can be offset is 0 bytes
```

W/ the help of @musjj's own minimal example using `bevy_ecs_ldtk`, I've updated the `basic` example of `bevy_ecs_tilemap` to encounter this bug in the following branch: https://github.com/Trouv/bevy_ecs_tilemap/blob/bug/0.12-render-target/examples/basic.rs#L173

In that system, waiting until frame 10 to spawn the camera causes the crash, but removing the `frame_count` restriction and having it spawn the camera on frame 0 fixes the crash.

**Comment by divark** [comment]
@Trouv Thanks for the heads up and the augmented `basic` example! With closer inspection, it looks like I missed a spot when performing the RenderSet changes. The give away was from that first error listed where it referenced the "tilemap_view_bind_group", which originates from the function `queue_material_tilemap_meshes` that tries to create those bind groups. That function was not registered in the new RenderSet `PrepareBindGroups`, which is necessary for Bind Group creation.

Luckily, this is a one-line change, and it seems that example now works on my end. I'll push this new commit now. Let me know if this change resolves the issue, as well as whether all the other examples still work as intended!

**Comment by Trouv** [comment]
I can confirm this does seem to fix the crash, but I am also running into new rendering issues now. In particular sprites seem to be rendering behind tilemaps, and the ordering of the tilemaps themselves seems to be inconsistent
On 50718d6ec60ff29abc0496446a4a1614d1a779fe:
platformer-example-50718d6ec60ff29abc0496446a4a1614d1a779fe

On dbc6b17365434bd0194d82a5426ce0561829e628:
platformer-example-dbc6b17365434bd0194d82a5426ce0561829e628

It seems to be some regression with z-ordering somehow.

**Comment by Data5tream** [comment]
@divark I was having the same issue and just tried your new commit. It seems to fix the z-index issue, but I sometimes get some weird rendering (happens only sometimes when I start my app, also happend with the previous version):
| Normal rendering (most times) | Weird rendering (few times) |
| --- | --- |
| old_normal | old_weird_rendering |
| new_normal | new_weird |

**Comment by divark** [comment]
> @divark I was having the same issue and just tried your new commit. It seems to fix the z-index issue, but I sometimes get some weird rendering (happens only sometimes when I start my app, also happend with the previous version):

When you say previous version, do you mean the last commit, or v0.11.1, the last version of this library?

**Comment by Trouv** [comment]
For what it's worth, things seem to be working correctly for `bevy_ecs_ldtk` again - including the late RenderTargets.

**Comment by musjj** [comment]
Just tested it and everything now works perfectly for me with `bevy_ecs_ldtk`, thanks for the fix!

**Comment by MickHarrigan** [comment]
> can we get this shipped so we can use it in the jam please 🥺

Quick update for any that are looking to use this in the jam.

To use this with `bevy_ecs_ldtk` add
```
[patch.crates-io]
bevy_ecs_tilemap = { git = "https://github.com/divark/bevy_ecs_tilemap", branch = "0.12-fixes" }
```
to your Cargo.toml, as well as 
```
bevy_ecs_ldtk = { git = "https://github.com/Trouv/bevy_ecs_ldtk.git", branch = "feat/bevy-0.12" }
```

These will make it that it uses the branches that are capable of using bevy 0.12 now, before they have been put on crates.io or the main branches of the respective repos. Hope this helped!

**Comment by mbrc12** [comment]
Maybe I am wrong, but there appears to be some issues with the Tiled helper. Consider this minimal example https://github.com/mbrc12/minimal-example-for-tiled-issue . The file `src/tiled.rs` is copied from `examples/helpers/tiled.rs`.

Specifically, the issue appears to be that the `tiled` crate processes the image path as described here: https://docs.rs/tiled/latest/tiled/struct.Image.html, which is different from what is assumed by the current helper code. A potential fix is in `tiled_fix.rs` which works for my case, but may fail elsewhere.

**Comment by divark** [comment]
> Specifically, the issue appears to be that the `tiled` crate processes the image path as described here: https://docs.rs/tiled/latest/tiled/struct.Image.html, which is different from what is assumed by the current helper code. A potential fix is in `tiled_fix.rs` which works for my case, but may fail elsewhere.

The image path from the Tiled library processes an uncanonicalized path, meaning the paths are relative, not absolute, which could cause problems as noted here: #484. I believe the Tiled helper is designed for absolute/canonical paths in mind, hence the extra setup to get the path from the asset loader and such.

If you think this warrants further investigation, I'd recommend making a new issue pointing it out unless this is a regression in comparison to v0.11.1 of this library.

**Pull request closed without merging**